src/preferences/effectmanifesttablemodel.cpp

```diff
@@ -0,0 +1,155 @@
+    return QAbstractTableModel::flags(index) | Qt::ItemIsSelectable |
+            Qt::ItemIsDragEnabled | Qt::ItemIsDropEnabled;
+}
+
+QMimeData* EffectManifestTableModel::mimeData(
+        const QModelIndexList& indexes) const {
+    QMimeData* mimeData = new QMimeData();
+    QStringList manifestUniqueIds;
+
+    for (const auto& index : indexes) {
+        if (!index.isValid() || index.row() < 0 ||
+                index.row() >= m_manifests.size()
+                // QModelIndexes for both columns are dragged, but only one of them in each row is needed.
+                || index.column() != 0) {
+            continue;
+        }
+        EffectManifestPointer pManifest = m_manifests.at(index.row());
+        manifestUniqueIds << pManifest->uniqueId();
+    }
+    mimeData->setText(manifestUniqueIds.join(kMimeTextDelimiter));
+    return mimeData;
+}
+
+bool EffectManifestTableModel::dropMimeData(const QMimeData* data,
+        Qt::DropAction action,
+        int row,
+        int column,
+        const QModelIndex& parent) {
+    Q_UNUSED(column);
+    if (!data->hasText() || action != Qt::MoveAction) {
+        return false;
+    }
+    QStringList mimeTextLines =
+#if QT_VERSION >= QT_VERSION_CHECK(5, 14, 0)
+            data->text().split(kMimeTextDelimiter, Qt::SkipEmptyParts);
+#else
+            data->text().split(kMimeTextDelimiter, QString::SkipEmptyParts);
+#endif
+    QList<EffectManifestPointer> manifestList;
+    for (int lineNumber = 0; lineNumber < mimeTextLines.size(); lineNumber++) {
+        QString manifestUniqueId = mimeTextLines.at(lineNumber);
+        EffectManifestPointer pManifest =
+                m_pBackendManager->getManifestFromUniqueId(manifestUniqueId);
+        // Do not VERIFY_OR_DEBUG_ASSERT here because MIME data could come from anywhere.
+        if (pManifest) {
+            manifestList.append(pManifest);
+        }
+    }
+    if (manifestList.isEmpty()) {
+        return false;
+    }
+    if (row == -1) {
+        row = parent.row();
+        // Dropping onto an empty model or dropping past the end of a model
+        if (parent.row() == -1) {
+            row = m_manifests.size();
+        }
+    }
+    beginInsertRows(QModelIndex(), row, row + manifestList.size() - 1);
+    for (int manifestNumber = 0; manifestNumber < manifestList.size(); manifestNumber++) {
+        m_manifests.insert(row + manifestNumber, manifestList.at(manifestNumber));
+    }
+    endInsertRows();
+    return true;
+}
+
+QStringList EffectManifestTableModel::mimeTypes() const {
+    return kAcceptedMimeTypes;
+}
+
+Qt::DropActions EffectManifestTableModel::supportedDropActions() const {
+    return Qt::MoveAction;
+}
+
+bool EffectManifestTableModel::removeRows(int row, int count, const QModelIndex& parent) {
+    Q_UNUSED(parent);
+    beginRemoveRows(QModelIndex(), row, row + count - 1);
+    for (int i = row; i < row + count; i++) {
+        m_manifests.removeAt(i);
```

Comment:
When I open the preferences and then press "Cancel", I get this warning:

```
warning [Main] QList::removeAt(): Index out of range.
```

Backtrace:
```
#0  0x00007ffff46eb615 in raise () at /usr/lib/libc.so.6
#1  0x00007ffff46d4862 in abort () at /usr/lib/libc.so.6
#2  0x00007ffff4cbb228 in QMessageLogger::warning(char const*, ...) const () at /usr/lib/libQt5Core.so.5
#3  0x00005555556a102b in QList<QSharedPointer<EffectManifest> >::removeAt(int) (this=<optimized out>, i=<optimized out>) at /usr/include/qt/QtCore/qlogging.h:90
#4  QList<QSharedPointer<EffectManifest> >::removeAt(int) (i=<optimized out>, this=<optimized out>) at /usr/include/qt/QtCore/qlist.h:582
#5  EffectManifestTableModel::removeRows(int, int, QModelIndex const&) (this=0x55555a37d090, row=<optimized out>, count=<optimized out>, parent=...) at /home/jan/Projects/mixxx/src/preferences/effectmanifesttablemodel.cpp:151
#6  0x0000555555b5d551 in EffectManifestTableModel::setList(QList<QSharedPointer<EffectManifest> > const&) (this=0x55555a37d090, newList=...) at /usr/include/qt/QtCore/qlist.h:176
#7  0x0000555555afea47 in DlgPrefEffects::slotUpdate() (this=0x55555a2f0940) at /home/jan/Projects/mixxx/src/preferences/dialog/dlgprefeffects.cpp:80
#8  0x00007ffff4f16070 in  () at /usr/lib/libQt5Core.so.5
#9  0x0000555555b122b0 in DlgPreferences::slotButtonPressed(QAbstractButton*) (this=0x5555592dcc70, pButton=<optimized out>) at /home/jan/Projects/mixxx/src/preferences/dialog/dlgpreferences.cpp:386
#10 0x00007ffff4f16070 in  () at /usr/lib/libQt5Core.so.5
#11 0x00007ffff69dbb40 in QDialogButtonBox::clicked(QAbstractButton*) () at /usr/lib/libQt5Widgets.so.5
#12 0x00007ffff69dc220 in  () at /usr/lib/libQt5Widgets.so.5
#13 0x00007ffff4f16070 in  () at /usr/lib/libQt5Core.so.5
#14 0x00007ffff6935b23 in QAbstractButton::clicked(bool) () at /usr/lib/libQt5Widgets.so.5
#15 0x00007ffff69364ac in  () at /usr/lib/libQt5Widgets.so.5
#16 0x00007ffff6937e53 in  () at /usr/lib/libQt5Widgets.so.5
#17 0x00007ffff6938033 in QAbstractButton::mouseReleaseEvent(QMouseEvent*) () at /usr/lib/libQt5Widgets.so.5
#18 0x00007ffff6885b0e in QWidget::event(QEvent*) () at /usr/lib/libQt5Widgets.so.5
#19 0x00007ffff6844752 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () at /usr/lib/libQt5Widgets.so.5
#20 0x00007ffff684b87b in QApplication::notify(QObject*, QEvent*) () at /usr/lib/libQt5Widgets.so.5
#21 0x00007ffff4edecda in QCoreApplication::notifyInternal2(QObject*, QEvent*) () at /usr/lib/libQt5Core.so.5
#22 0x00007ffff684a87e in QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool, bool) () at /usr/lib/libQt5Widgets.so.5
#23 0x00007ffff689e249 in  () at /usr/lib/libQt5Widgets.so.5
#24 0x00007ffff68a163f in  () at /usr/lib/libQt5Widgets.so.5
#25 0x00007ffff6844752 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () at /usr/lib/libQt5Widgets.so.5
#26 0x00007ffff4edecda in QCoreApplication::notifyInternal2(QObject*, QEvent*) () at /usr/lib/libQt5Core.so.5
#27 0x00007ffff5f8e4ac in QGuiApplicationPrivate::processMouseEvent(QWindowSystemInterfacePrivate::MouseEvent*) () at /usr/lib/libQt5Gui.so.5
#28 0x00007ffff5f63bac in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /usr/lib/libQt5Gui.so.5
#29 0x00007fffed2a0f9c in  () at /usr/lib/libQt5XcbQpa.so.5
#30 0x00007ffff5ba3914 in g_main_context_dispatch () at /usr/lib/libglib-2.0.so.0
#31 0x00007ffff5bf77d1 in  () at /usr/lib/libglib-2.0.so.0
#32 0x00007ffff5ba2121 in g_main_context_iteration () at /usr/lib/libglib-2.0.so.0
#33 0x00007ffff4f37941 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /usr/lib/libQt5Core.so.5
#34 0x00007ffff4edd65c in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () at /usr/lib/libQt5Core.so.5
#35 0x00007ffff4ee5af4 in QCoreApplication::exec() () at /usr/lib/libQt5Core.so.5
#36 0x00005555556f8df6 in (anonymous namespace)::runMixxx (args=..., app=0x7fffffffdb30) at /home/jan/Projects/mixxx/src/main.cpp:53
#37 main(int, char**) (argc=<optimized out>, argv=<optimized out>) at /home/jan/Projects/mixxx/src/main.cpp:133
```

Reply:
The same happens if I select some LV2 effects in the "Visible Effects" tab and drag-and-drop it into the left list:

```
#0  0x00007ffff46eb615 in raise () at /usr/lib/libc.so.6
#1  0x00007ffff46d4862 in abort () at /usr/lib/libc.so.6
#2  0x00007ffff4cbb228 in QMessageLogger::warning(char const*, ...) const () at /usr/lib/libQt5Core.so.5
#3  0x00005555556a102b in QList<QSharedPointer<EffectManifest> >::removeAt(int) (this=<optimized out>, i=<optimized out>) at /usr/include/qt/QtCore/qlogging.h:90
#4  QList<QSharedPointer<EffectManifest> >::removeAt(int) (i=<optimized out>, this=<optimized out>) at /usr/include/qt/QtCore/qlist.h:582
#5  EffectManifestTableModel::removeRows(int, int, QModelIndex const&) (this=0x55555a3829d0, row=<optimized out>, count=<optimized out>, parent=...) at /home/jan/Projects/mixxx/src/preferences/effectmanifesttablemodel.cpp:151
#6  0x00007ffff6ab2ffc in  () at /usr/lib/libQt5Widgets.so.5
#7  0x00007ffff6ab5198 in QAbstractItemView::startDrag(QFlags<Qt::DropAction>) () at /usr/lib/libQt5Widgets.so.5
#8  0x00007ffff6ab39a0 in QAbstractItemView::mouseMoveEvent(QMouseEvent*) () at /usr/lib/libQt5Widgets.so.5
#9  0x00007ffff6885b0e in QWidget::event(QEvent*) () at /usr/lib/libQt5Widgets.so.5
#10 0x00007ffff6930f3f in QFrame::event(QEvent*) () at /usr/lib/libQt5Widgets.so.5
#11 0x00007ffff4edea43 in QCoreApplicationPrivate::sendThroughObjectEventFilters(QObject*, QEvent*) () at /usr/lib/libQt5Core.so.5
#12 0x00007ffff6844741 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () at /usr/lib/libQt5Widgets.so.5
#13 0x00007ffff684b87b in QApplication::notify(QObject*, QEvent*) () at /usr/lib/libQt5Widgets.so.5
#14 0x00007ffff4edecda in QCoreApplication::notifyInternal2(QObject*, QEvent*) () at /usr/lib/libQt5Core.so.5
#15 0x00007ffff684a87e in QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool, bool) () at /usr/lib/libQt5Widgets.so.5
#16 0x00007ffff689e249 in  () at /usr/lib/libQt5Widgets.so.5
#17 0x00007ffff68a163f in  () at /usr/lib/libQt5Widgets.so.5
#18 0x00007ffff6844752 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () at /usr/lib/libQt5Widgets.so.5
#19 0x00007ffff4edecda in QCoreApplication::notifyInternal2(QObject*, QEvent*) () at /usr/lib/libQt5Core.so.5
#20 0x00007ffff5f8e4ac in QGuiApplicationPrivate::processMouseEvent(QWindowSystemInterfacePrivate::MouseEvent*) () at /usr/lib/libQt5Gui.so.5
#21 0x00007ffff5f63bac in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /usr/lib/libQt5Gui.so.5
#22 0x00007fffed2a0f9c in  () at /usr/lib/libQt5XcbQpa.so.5
#23 0x00007ffff5ba3914 in g_main_context_dispatch () at /usr/lib/libglib-2.0.so.0
#24 0x00007ffff5bf77d1 in  () at /usr/lib/libglib-2.0.so.0
#25 0x00007ffff5ba2121 in g_main_context_iteration () at /usr/lib/libglib-2.0.so.0
#26 0x00007ffff4f37941 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /usr/lib/libQt5Core.so.5
#27 0x00007ffff4edd65c in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () at /usr/lib/libQt5Core.so.5
#28 0x00007ffff4ee5af4 in QCoreApplication::exec() () at /usr/lib/libQt5Core.so.5
#29 0x00005555556f8df6 in (anonymous namespace)::runMixxx (args=..., app=0x7fffffffdb30) at /home/jan/Projects/mixxx/src/main.cpp:53
#30 main(int, char**) (argc=<optimized out>, argv=<optimized out>) at /home/jan/Projects/mixxx/src/main.cpp:133
```

Reply:
I can reproduce this since upgrading to Fedora 33 with Qt 5.15.1. I'll look into it.

Reply:
fixed in 55041457b8c2efa8e17fc3ba2559a35cfcb82cf1 ... that was puzzling